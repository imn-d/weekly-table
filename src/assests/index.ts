export * from './CancelIcon';
export * from './DeleteIcon';
export * from './EditIcon';
export * from './DragIcon';
export * from './CheckIcon';
