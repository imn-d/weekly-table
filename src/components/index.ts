export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
