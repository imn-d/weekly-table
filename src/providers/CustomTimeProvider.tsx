import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useState,
} from 'react';

import {
  CustomTimeProps,
  DisplayTime,
  HOUR_24,
  Position,
  TIME,
  Time,
  TimeBlock,
} from '../common';
import { useCells } from './CellsProvider';
import { useTimeBlock } from './TimeBlockProvider';

const CustomTimeContext = createContext<CustomTimeProps>({} as CustomTimeProps);

/**
 *  Custom time provider context hook
 *  @returns [CustomTimeProps]{@link CustomTimeProps}
 */
export const useCustomTime = (): CustomTimeProps =>
  useContext(CustomTimeContext);

/**
 * Custom time provider, used for setting time with second precision
 */
const CustomTimeProvider: FC = ({ children }) => {
  const { millisPerPixel } = useCells();
  const { blocks, setBlocks } = useTimeBlock();

  /**
   * Ms time state
   */
  const [time, setTime] = useState<Time>({} as Time);

  /**
   * Timers relative coordinates
   */
  const [position, setPosition] = useState<Position>();

  /**
   * Editing state
   * If active, other operations with timeblocks will be ignored
   */
  const [isEditing, setEditing] = useState<boolean>(false);

  /**
   * Prime timeblock state
   */
  const [oldBlock, setOldBlock] = useState<TimeBlock>();

  /**
   * Error state with description
   */
  const [isError, setError] = useState<string | undefined>(undefined);

  /**
   * Timers visible time
   */
  const [displayTime, setDisplayTime] = useState<DisplayTime>({
    startTime: '00:00:00',
    endTime: '00:00:00',
  } as DisplayTime);

  /**
   * Extracting timeblock properties and saving it
   */
  const handleObject = useCallback((block: TimeBlock) => {
    setOldBlock(block);
    setPosition({
      top: block.height > 20 ? block.height + block.top + 5 : block.top + 25,
      left: block.left,
    });
    setTime({
      startTime: block.startTime,
      endTime: block.endTime,
    });
    setDisplayTime({
      startTime: new Date(block.startTime).toISOString().substr(11, 8),
      endTime: new Date(block.endTime).toISOString().substr(11, 8),
    });
  }, []);

  /**
   * Timers data extraction
   */
  const handleTime = useCallback(
    (newTime: number, type: boolean) => {
      if (newTime === undefined || newTime === null) {
        setError('Time is null');
        return;
      }
      if (isNaN(newTime)) {
        newTime = 0;
      }
      setError(undefined);
      if (type === TIME.START) {
        setTime({ ...time, startTime: newTime });
        setDisplayTime({
          ...displayTime,
          startTime: new Date(newTime).toISOString().substr(11, 8),
        });
      } else {
        setTime({ ...time, endTime: newTime === 0 ? HOUR_24 : newTime });
        setDisplayTime({
          ...displayTime,
          endTime: new Date(newTime).toISOString().substr(11, 8),
        });
      }
    },
    [displayTime, time],
  );

  /**
   * Timers data reset
   */
  const resetTime = useCallback(() => {
    if (!oldBlock) return;
    setTime({
      startTime: oldBlock.startTime,
      endTime: oldBlock.endTime,
    });
    setDisplayTime({
      startTime: new Date(oldBlock.startTime).toISOString().substr(11, 8),
      endTime: new Date(oldBlock.endTime).toISOString().substr(11, 8),
    });
    setError(undefined);
  }, [oldBlock]);

  /**
   * Updating time to timeblock
   */
  const updateBlock = useCallback(() => {
    if (!oldBlock) return true;
    const filteredBlocks = blocks.filter((b) => b.id !== oldBlock.id);

    if (time.endTime - time.startTime < 1000 * 60) {
      setError('Range < 1 min');
      return true;
    }
    const topY = -(oldBlock.startTime - time.startTime) / millisPerPixel;
    const botY = -(oldBlock.endTime - time.endTime) / millisPerPixel;

    const newBlock = {
      ...oldBlock,
      startTime: time.startTime,
      endTime: time.endTime,
      realStartTime: time.startTime,
      realEndTime: time.endTime,
      top: oldBlock.top + topY,
      height: oldBlock.height + botY - topY,
    };
    setBlocks(filteredBlocks.concat(newBlock));
    return false;
    //eslint-disable-next-line
  }, [blocks, millisPerPixel, oldBlock, time.endTime, time.startTime]);

  return (
    <CustomTimeContext.Provider
      value={{
        isEditing,
        resetTime,
        setEditing,
        handleTime,
        updateBlock,
        position,
        handleObject,
        displayTime,
        isError,
      }}
    >
      {children}
    </CustomTimeContext.Provider>
  );
};

export default CustomTimeProvider;
