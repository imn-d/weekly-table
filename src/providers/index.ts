export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
