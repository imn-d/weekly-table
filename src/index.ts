import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
