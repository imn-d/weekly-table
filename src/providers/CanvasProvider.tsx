import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';

import { CanvasProps } from '../common';
import { useCells } from './CellsProvider';
import { useScheduler } from './Scheduler';

const CanvasContext = createContext<CanvasProps>({} as CanvasProps);
/**
 *  Drawer provider context hook
 *  @returns [CanvasProps]{@link CanvasProps}
 */
export const useCanvas = (): CanvasProps => useContext(CanvasContext);

/**
 * Drawing provider, used for timeblocks creation
 */
const CanvasProvider: FC = ({ children }) => {
  const { headerHeightProp, helperWidthProp, bottomHeightProp } =
    useScheduler();
  const { offsetWidth, offsetHeight, cells, setCells } = useCells();

  /**
   *  Drawing state
   */
  const [isDrawing, setIsDrawing] = useState<boolean>(false);

  /**
   * Canvas reference
   */
  const canvasRef = useRef<HTMLCanvasElement>(null);

  /**
   * Context reference
   */
  const contextRef = useRef<CanvasRenderingContext2D | null>(null);

  /**
   * Start point relative coordinates
   */
  const [start, setStart] = useState<[number, number]>([0, 0]);

  /**
   * End point relative coordinates
   */
  const [end, setEnd] = useState<[number, number] | undefined>(undefined);

  /**
   * Current position relative coordinates
   */
  const [current, setCurrent] = useState<[number, number] | undefined>(
    undefined,
  );

  /**
   * Drawing area init
   */
  useEffect(() => {
    const canvas = canvasRef.current;
    if (!canvas) return;
    canvas.width = offsetWidth * 2;
    canvas.height = offsetHeight * 2;
    canvas.style.width = `${offsetWidth}px`;
    canvas.style.height = `${offsetHeight}px`;
    canvas.draggable = false;

    const context = canvas.getContext('2d');
    if (!context) return;
    context.scale(2, 2);
    context.lineCap = 'round';
    context.strokeStyle = 'blue';
    context.lineWidth = 1;
    context.fillStyle = '#0000FF19';
    contextRef.current = context;
  }, [offsetHeight, offsetWidth]);

  /**
   * Start drawing callback
   */
  const startDrawing = useCallback(({ nativeEvent }: React.MouseEvent) => {
    const { offsetX, offsetY } = nativeEvent;
    if (!contextRef.current) return;
    contextRef.current.beginPath();
    contextRef.current.moveTo(offsetX, offsetY);
    setIsDrawing(true);
    setStart([offsetX, offsetY]);
    setEnd(undefined);
  }, []);

  /**
   * Clearing canvas callback
   */
  const clearCanvas = useCallback(() => {
    if (!contextRef.current) {
      return;
    }
    contextRef.current.clearRect(0, 0, offsetWidth, offsetHeight);
  }, [offsetHeight, offsetWidth]);

  /**
   * End drawing callback
   */
  const finishDrawing = useCallback(
    ({ nativeEvent }: React.MouseEvent) => {
      if (!contextRef.current) return;
      if (!isDrawing) return;

      contextRef.current.closePath();
      setIsDrawing(false);
      setEnd([nativeEvent.offsetX, nativeEvent.offsetY]);
      setCurrent(undefined);
      clearCanvas();
    },
    //eslint-disable-next-line
    [isDrawing],
  );

  /**
   * Drawing process callback
   */
  const draw = useCallback(
    ({ nativeEvent }: React.MouseEvent) => {
      if (!isDrawing || !contextRef.current) {
        return;
      }
      setCurrent([nativeEvent.offsetX, nativeEvent.offsetY]);
      contextRef.current.clearRect(0, 0, offsetWidth, offsetHeight);
      contextRef.current.strokeRect(
        start[0],
        start[1],
        nativeEvent.offsetX - start[0],
        nativeEvent.offsetY - start[1],
      );
      contextRef.current.fillRect(
        start[0] + 1,
        start[1] + 1,
        nativeEvent.offsetX - start[0] - 1,
        nativeEvent.offsetY - start[1] - 1,
      );
    },
    [isDrawing, offsetHeight, offsetWidth, start],
  );

  /**
   * Finding cell index by coordinates
   */
  const findCell = useCallback(
    (position: [number, number] | undefined): number => {
      if (!position || !cells) return -1;
      if (position[0] >= offsetWidth - 5) position[0] = offsetWidth - 5;
      if (position[0] <= helperWidthProp) position[0] = helperWidthProp;
      if (position[1] <= headerHeightProp) position[1] = headerHeightProp;
      if (position[1] >= offsetHeight - bottomHeightProp)
        position[1] = offsetHeight - bottomHeightProp - 1;

      return cells.findIndex(
        (cell) =>
          cell.position.x <= position[0] &&
          position[0] <= cell.position.w &&
          cell.position.y <= position[1] &&
          position[1] <= cell.position.h,
      );
    },
    [
      offsetWidth,
      cells,
      helperWidthProp,
      headerHeightProp,
      offsetHeight,
      bottomHeightProp,
    ],
  );

  /**
   * Start, end, current coordinates changes watcher
   * Calculates selected cells area and set isSelected property to it
   */
  useEffect(() => {
    if (end === undefined && current === undefined) return;
    const startCell = findCell(start);
    if (startCell === -1) return;
    const endCell = findCell(current === undefined ? end : current);
    if (endCell === -1) return;
    const sCell = cells[startCell];
    const eCell = cells[endCell];
    const sRow = sCell.row < eCell.row ? sCell.row : eCell.row;
    const sCol = sCell.column < eCell.column ? sCell.column : eCell.column;
    const eRow = eCell.row > sCell.row ? eCell.row : sCell.row;
    const eCol = eCell.column > sCell.column ? eCell.column : sCell.column;

    const insertCells = cells.map((cell) => {
      if (
        cell.row >= sRow &&
        cell.row <= eRow &&
        cell.column >= sCol &&
        cell.column <= eCol
      ) {
        return { ...cell, isSelected: true };
      } else return cell;
    });
    setCells(insertCells);
    //eslint-disable-next-line
  }, [start, end, current]);

  return (
    <CanvasContext.Provider
      value={{
        canvasRef,
        startDrawing,
        finishDrawing,
        draw,
        isDrawing,
      }}
    >
      {children}
    </CanvasContext.Provider>
  );
};

export default CanvasProvider;
