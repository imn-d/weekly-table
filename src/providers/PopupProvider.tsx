import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import {
  ACTION,
  DIR,
  HOUR_24,
  PopupProps,
  PopupType,
  TimeBlock,
} from '../common';
import { useCells } from './CellsProvider';
import { usePointerLock } from './PointerLockProvider';
import { useScheduler } from './Scheduler';
import { useTimeBlock } from './TimeBlockProvider';

const PopupContext = createContext<PopupProps>({} as PopupProps);

/**
 *  Popup provider context hook
 *  @returns [PopupProps]{@link PopupProps}
 */
export const usePopup = (): PopupProps => useContext(PopupContext);

/**
 * Popup provider, operates most timeblock actions
 */
const PopupProvider: FC = ({ children }) => {
  const { blocks, setBlocks, handleHistory, undoHistory } = useTimeBlock();
  const { millisPerPixel, msTime } = useCells();
  const { isLocking, movement, action } = usePointerLock();
  const { mouseSpeed } = useScheduler();
  /**
   * Popup state, component have only one popup
   */
  const [popup, setPopup] = useState<PopupType | undefined>(undefined);

  /**
   * Hotkeys state
   */
  const [keyRequest, setKeyRequest] = useState({ code: '', ctrl: false });

  /**
   * Hotkeys event listeners
   */
  useEffect(() => {
    document.addEventListener('keydown', handleKey);
    return () => {
      document.removeEventListener('keydown', handleKey);
    };
    //eslint-disable-next-line
  }, []);

  /**
   * Hotkeys actions handler
   */
  const handleKey = useCallback((event: KeyboardEvent) => {
    if (event.code) {
      setKeyRequest({ code: event.code, ctrl: event.ctrlKey });
    }
  }, []);

  /**
   * Hotkeys actions
   */
  useEffect(() => {
    if (keyRequest.code === 'Delete' || keyRequest.code === 'Backspace') {
      if (popup?.id) {
        deleteBlock(popup.id);
      }
      setKeyRequest({ code: '', ctrl: false });
      return;
    }
    if (keyRequest.code === 'KeyZ' && keyRequest.ctrl) {
      undoHistory();
      setKeyRequest({ code: '', ctrl: false });
    }
    //eslint-disable-next-line
  }, [keyRequest.code, keyRequest.ctrl, popup?.id]);

  /**
   * Popup rendering over timeblock
   */
  const showPopup = useCallback(
    (blockId: string) => {
      const block = blocks.find((b) => b.id === blockId);
      if (!block) return;
      setPopup({
        id: block.id,
        block: block,
      });
    },
    [blocks],
  );

  /**
   * Popup hiding eq removing
   */
  const hidePopup = useCallback(() => {
    setPopup(undefined);
  }, []);

  /**
   * isAllowing timerange state
   * Timeblock will stopping actions if it less than timeframe
   * Don't affecting to custom time setting
   */
  const isAllowing = useMemo<boolean>(() => {
    if (!popup?.block) return true;
    return popup.block.endTime - popup.block.startTime >= msTime;
  }, [msTime, popup?.block]);

  /**
   * Block remover
   */
  const deleteBlock = useCallback(
    (blockId: string) => {
      setBlocks([...blocks].filter((b) => b.id !== blockId));
      handleHistory([...blocks].filter((b) => b.id !== blockId));
      setPopup(undefined);
    },
    //eslint-disable-next-line
    [blocks],
  );

  /**
   * Move handler - moving block with mouse pressed
   * Hidden realTime field used
   * Rounding visible time every timeframe
   */
  const handleMove = useCallback(
    (block: TimeBlock, movementY: number) => {
      const oldTop = block.top;
      const newTop = oldTop + movementY * mouseSpeed;

      const newBlocks = [...blocks].filter((b) => b.id !== block.id);
      const dateDiff = (newTop - oldTop) * millisPerPixel;
      const newStart = block.realStartTime + dateDiff;
      const newEnd = block.realEndTime + dateDiff;
      if (
        newStart < -msTime / 3 ||
        newEnd > HOUR_24 + msTime / 3 ||
        newEnd - newStart < msTime
      ) {
        return;
      }

      const roundedStart = Math.abs(
        Math.round(block.realStartTime / msTime) * msTime,
      );
      const roundedEnd = Math.round(block.realEndTime / msTime) * msTime;

      const movedBlock = {
        ...block,
        startTime: roundedStart,
        endTime: roundedEnd,
        realStartTime: newStart,
        realEndTime: newEnd,
        top: newTop,
      };
      setBlocks(newBlocks.concat(movedBlock));
    },
    [mouseSpeed, blocks, millisPerPixel, msTime],
  );

  /**
   * Resize handler - resizing block with mouse pressed
   * Hidden realTime field used
   * Rounding visible time every timeframe
   */
  const handleResize = useCallback(
    (block: TimeBlock, movementY: number, dir: boolean) => {
      const newBlocks = blocks.filter((b) => b.id !== block.id);
      if (dir === DIR.TOP) {
        const oldTop = block.top;
        const newTop = oldTop + movementY * mouseSpeed;
        const dateDiff = (newTop - oldTop) * millisPerPixel;
        const newStart = block.realStartTime + dateDiff;

        if (newStart < -msTime / 3) return;
        if (block.realEndTime - newStart < msTime) return;

        const roundedStart = Math.abs(
          Math.round(block.realStartTime / msTime) * msTime,
        );

        const movedBlock = {
          ...block,
          startTime: roundedStart,
          realStartTime: newStart,
          top: newTop,
          height: block.height + oldTop - newTop,
        };
        setBlocks(newBlocks.concat(movedBlock));
        return;
      }

      if (dir === DIR.BOTTOM) {
        const oldBot = block.height;
        const newBot = oldBot + movementY * mouseSpeed;
        const dateDiff = (newBot - oldBot) * millisPerPixel;

        const newEnd = block.realEndTime + dateDiff;
        if (newEnd > HOUR_24 + msTime / 3) return;
        if (newEnd - block.realStartTime < msTime) return;

        let roundedEnd = Math.round(block.realEndTime / msTime) * msTime;
        if (roundedEnd > HOUR_24) roundedEnd = HOUR_24;

        const movedBlock = {
          ...block,
          endTime: roundedEnd,
          realEndTime: newEnd,
          height: newBot,
        };
        setBlocks(newBlocks.concat(movedBlock));
      }
    },
    [blocks, millisPerPixel, mouseSpeed, msTime],
  );

  /**
   * mouseevents router when timeblock actions
   */
  useEffect(() => {
    if (isLocking && popup?.block && movement !== 0) {
      switch (action) {
        case ACTION.MOVE: {
          handleMove(popup.block, movement);
          return;
        }
        case ACTION.RESIZE_TOP: {
          handleResize(popup.block, movement, DIR.TOP);
          return;
        }
        case ACTION.RESIZE_BOT: {
          handleResize(popup.block, movement, DIR.BOTTOM);
          return;
        }
        default:
          return;
      }
    }
  }, [action, isLocking, movement, popup?.block]);

  return (
    <PopupContext.Provider
      value={{
        isAllowing,
        popup,
        hidePopup,
        deleteBlock,
        showPopup,
      }}
    >
      {children}
    </PopupContext.Provider>
  );
};

export default PopupProvider;
