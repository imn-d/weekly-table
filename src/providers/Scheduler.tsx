import React, {
  FC,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react';

import {
  BlockColorsProps,
  SchedulerInputProps,
  SchedulerProps,
  schedulerColumns,
  schedulerRows,
} from '../common';
import { Background, CustomTime, Popup, SchedulerLayout } from '../components';
import CanvasProvider from './CanvasProvider';
import CellsProvider from './CellsProvider';
import CustomTimeProvider from './CustomTimeProvider';
import PointerLockProvider from './PointerLockProvider';
import PopupProvider from './PopupProvider';
import TimeBlockProvider from './TimeBlockProvider';

const SchedulerContext = createContext<SchedulerProps>({} as SchedulerProps);
/**
 *  Scheduler provider context hook
 *  @returns [SchedulerProps]{@link SchedulerProps}
 */
export const useScheduler = (): SchedulerProps => useContext(SchedulerContext);

/**
 * Root scheduler components
 * @param props{@link SchedulerInputProps} initial props
 */
const Scheduler: FC<SchedulerInputProps> = (props: SchedulerInputProps) => {
  const [update, setUpdate] = useState<number>(0);

  useEffect(() => {
    window.addEventListener('resize', () => setUpdate((update) => update + 1));
    return () => {
      window.removeEventListener('resize', () =>
        setUpdate((update) => update + 1),
      );
    };
  }, []);

  return (
    <SchedulerContext.Provider
      value={{
        parentRef: props.parentRef,
        blockColors: props?.blockColors
          ? props.blockColors
          : ({
              common: '#ff5722',
              temp: '#c6a700',
              draw: '#ff8a50',
              hover: '#ff3d00',
            } as BlockColorsProps),
        headerHeightProp: props?.headerHeightProp ? props.headerHeightProp : 80,
        helperWidthProp: props?.helperWidthProp ? props.helperWidthProp : 80,
        bottomHeightProp: props?.bottomHeightProp ? props.bottomHeightProp : 20,
        baseZIndex: props?.baseZIndex ? props.baseZIndex : 0,
        timeframe: props?.timeframe ? props.timeframe : 60,
        columns: props?.columns ? props.columns : schedulerColumns,
        rows: props?.rows ? props.rows : schedulerRows,
        requiredTZOffset: props?.requiredTZOffset,
        mouseSpeed: props?.mouseSpeed ? props.mouseSpeed : 3,
        defaultValue: props?.defaultValue,
        onChange: props?.onChange,
      }}
    >
      <CellsProvider>
        <CanvasProvider>
          <PointerLockProvider>
            <TimeBlockProvider>
              <PopupProvider>
                <CustomTimeProvider>
                  <Background />
                  <SchedulerLayout />
                  <Popup />
                  <CustomTime />
                </CustomTimeProvider>
              </PopupProvider>
            </TimeBlockProvider>
          </PointerLockProvider>
        </CanvasProvider>
      </CellsProvider>
    </SchedulerContext.Provider>
  );
};

export default Scheduler;
