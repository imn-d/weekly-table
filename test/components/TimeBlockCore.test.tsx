import { mount } from 'enzyme';
import React from 'react';

import {
  BlockColorsProps,
  CanvasProps,
  CustomTimeProps,
  HOUR_24,
  PopupProps,
  PopupType,
  SchedulerProps,
  TimeBlock,
} from '../../src/common';
import { TimeBlockCore } from '../../src/components';
import * as uc from '../../src/providers/CanvasProvider';
import * as ut from '../../src/providers/CustomTimeProvider';
import * as up from '../../src/providers/PopupProvider';
import * as us from '../../src/providers/Scheduler';

describe('TimeBlock Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useTimeMock = jest.spyOn(ut, 'useCustomTime');
  const usePopupMock = jest.spyOn(up, 'usePopup');
  const useCanvasMock = jest.spyOn(uc, 'useCanvas');

  const timeBlock = {
    id: '1',
    isTemp: false,
    realStartTime: 100000,
    realEndTime: 200000,
    startTime: 100000,
    endTime: 200000,
    column: 1,
    top: 0,
    height: 100,
    left: 20,
    width: 50,
  } as TimeBlock;

  const schedulerMock = {
    baseZIndex: 0,
    blockColors: {
      hover: 'hover',
      common: 'common',
      temp: 'temp',
      draw: 'draw',
    } as BlockColorsProps,
  } as SchedulerProps;

  const canvasMock = {
    isDrawing: false,
  } as CanvasProps;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const popupMock = {
    popup: undefined,
    showPopup: jest.fn(),
    hidePopup: jest.fn(),
  } as PopupProps;

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const timeMock = {
    isEditing: false,
    handleObject: jest.fn(),
  } as CustomTimeProps;

  beforeEach(() => {
    jest.clearAllMocks();
    useSchedulerMock.mockReturnValue(schedulerMock);
  });

  it('Base Render', () => {
    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce(popupMock);
    useTimeMock.mockReturnValueOnce(timeMock);

    const wrapper = mount(<TimeBlockCore {...timeBlock} />);
    expect(wrapper).toMatchSnapshot('baseRender');

    expect(useSchedulerMock).toBeCalledTimes(1);
    expect(useCanvasMock).toBeCalledTimes(1);
    expect(usePopupMock).toBeCalledTimes(1);
    expect(useTimeMock).toBeCalledTimes(1);
    wrapper.unmount();
  });

  it('Block Color Test', () => {
    useTimeMock.mockReturnValue(timeMock);

    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce(popupMock);

    let wrapper = mount(<TimeBlockCore {...timeBlock} />);
    expect(
      wrapper.find('#timeBlock').hostNodes().props()?.style?.backgroundColor,
    ).toBe('common');
    wrapper.unmount();

    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce({
      ...popupMock,
      popup: { id: '1' } as PopupType,
    });

    wrapper = mount(<TimeBlockCore {...timeBlock} />);
    expect(
      wrapper.find('#timeBlock').hostNodes().props()?.style?.backgroundColor,
    ).toBe('hover');
    wrapper.unmount();

    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce(popupMock);

    wrapper = mount(<TimeBlockCore {...{ ...timeBlock, isTemp: true }} />);
    expect(
      wrapper.find('#timeBlock').hostNodes().props()?.style?.backgroundColor,
    ).toBe('temp');
    wrapper.unmount();

    useCanvasMock.mockReturnValueOnce({ ...canvasMock, isDrawing: true });
    usePopupMock.mockReturnValueOnce(popupMock);

    wrapper = mount(<TimeBlockCore {...timeBlock} />);
    expect(
      wrapper.find('#timeBlock').hostNodes().props()?.style?.backgroundColor,
    ).toBe('draw');
    wrapper.unmount();
  });

  it('End Time Test', () => {
    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce(popupMock);
    useTimeMock.mockReturnValueOnce(timeMock);

    const wrapper = mount(
      <TimeBlockCore {...{ ...timeBlock, endTime: HOUR_24 }} />,
    );
    expect(wrapper.find('#timeBlock').hostNodes()).toMatchSnapshot(
      'display24Hour',
    );
    wrapper.unmount();
  });

  it('UX Test', () => {
    const showPopup = jest.fn();
    const handleObject = jest.fn();
    const hidePopup = jest.fn();

    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce({ ...popupMock, showPopup, hidePopup });
    useTimeMock.mockReturnValueOnce({ ...timeMock, handleObject });

    const wrapper = mount(<TimeBlockCore {...timeBlock} />);
    const block = wrapper.find('#timeBlock').hostNodes();

    expect(showPopup).toBeCalledTimes(0);
    expect(handleObject).toBeCalledTimes(0);
    expect(hidePopup).toBeCalledTimes(0);

    block.simulate('mouseenter');
    expect(showPopup).toBeCalledTimes(1);
    expect(handleObject).toBeCalledTimes(1);
    expect(hidePopup).toBeCalledTimes(0);

    block.simulate('mousemove');
    expect(showPopup).toBeCalledTimes(2);
    expect(handleObject).toBeCalledTimes(2);
    expect(hidePopup).toBeCalledTimes(0);

    block.simulate('mouseleave');
    expect(showPopup).toBeCalledTimes(2);
    expect(handleObject).toBeCalledTimes(2);
    expect(hidePopup).toBeCalledTimes(1);

    wrapper.unmount();
  });

  it('isEditing Test', () => {
    const showPopup = jest.fn();
    const handleObject = jest.fn();
    const hidePopup = jest.fn();

    useCanvasMock.mockReturnValueOnce(canvasMock);
    usePopupMock.mockReturnValueOnce({ ...popupMock, showPopup, hidePopup });
    useTimeMock.mockReturnValueOnce({
      ...timeMock,
      handleObject,
      isEditing: true,
    });

    const wrapper = mount(<TimeBlockCore {...timeBlock} />);
    const block = wrapper.find('#timeBlock').hostNodes();

    block.simulate('mouseleave');
    block.simulate('mouseenter');
    block.simulate('mousemove');

    expect(showPopup).toBeCalledTimes(0);
    expect(handleObject).toBeCalledTimes(0);
    expect(hidePopup).toBeCalledTimes(0);

    wrapper.unmount();
  });
});
