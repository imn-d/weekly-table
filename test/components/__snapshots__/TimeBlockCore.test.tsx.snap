// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeBlock Component Base Render: baseRender 1`] = `
<TimeBlockCore
  column={1}
  endTime={200000}
  height={100}
  id="1"
  isTemp={false}
  left={20}
  realEndTime={200000}
  realStartTime={100000}
  startTime={100000}
  top={0}
  width={50}
>
  <div
    className="scheduler-time-block"
    id="timeBlock"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "backgroundColor": "common",
        "height": 100,
        "left": 20,
        "position": "absolute",
        "top": 0,
        "width": 50,
        "zIndex": 4,
      }
    }
  >
    <span
      className="scheduler-text scheduler-time-display"
      id="timeDisplay"
      style={
        Object {
          "pointerEvents": "none",
          "userSelect": "none",
        }
      }
    >
      00:01 / 00:03
    </span>
  </div>
</TimeBlockCore>
`;

exports[`TimeBlock Component End Time Test: display24Hour 1`] = `
<div
  className="scheduler-time-block"
  id="timeBlock"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  style={
    Object {
      "backgroundColor": "common",
      "height": 100,
      "left": 20,
      "position": "absolute",
      "top": 0,
      "width": 50,
      "zIndex": 4,
    }
  }
>
  <span
    className="scheduler-text scheduler-time-display"
    id="timeDisplay"
    style={
      Object {
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
  >
    00:01 / 24:00
  </span>
</div>
`;
